packages/rum/src/domain/record/record.ts

```diff
@@ -40,14 +48,24 @@ export function record(options: RecordOptions): RecordAPI {
     throw new Error('emit function is required')
   }
 
-  const mutationController = new MutationController()
   const elementsScrollPositions = createElementsScrollPositions()
 
+  const mutationCb = (mutation: BrowserMutationPayload) => {
+    emit(assembleIncrementalSnapshot<BrowserMutationData>(IncrementalSource.Mutation, mutation))
+  }
+  const inputCb: InputCallback = (s) => emit(assembleIncrementalSnapshot<InputData>(IncrementalSource.Input, s))
+
+  const {
+    shadowDomCallBacks,
+    flush: flushMutationsFromShadowRoots,
+    shadowDomCreatedCallback,
+    shadowDomRemovedCallback,
+  } = withShadowDomHelpers(options.configuration, { mutationCb, inputCb })
```

Comment:
I've kept `shadowDomCallBacks` in there until we agree on the other comment here: https://github.com/DataDog/browser-sdk/pull/1787#discussion_r1047290624
Since it's leaking a lot of implementation, instead of exporting the Map we could export
```ts
interface ShadowRootsController {
  addShadowRoot: (shadowRoot: ShadowRoot) => void
  removeShadowRoot: (shadowRoot: ShadowRoot) => void
  stop: () => void
  flush: () => void
+  activeShadowRootCountRef: {
     value: number
  }
}
```

Reply:
`ShadowRootsController` looks good, we should still do something for `shadowDomCallBacks` though

Reply:
The `ShadowDomCallBacks` is an internal type only used by the `new Map<ShadowRoot, ShadowDomCallBacks>()`. What is the issue with it? What would you like to change?

Reply:
cf top of the thread:

> - shadowDomCallBacks is quite confusing since it is used for various different concepts:
>   - shadowDomCallBacks = new Map<ShadowRoot, ShadowDomCallBacks>()
>   - interface ShadowDomCallBacks (declared both in record.ts and shadowDom.ts)
>   - shadowDomCallBacks: { shadowDomCreatedCallback, shadowDomRemovedCallback }
>   - bonus: ShadowDomCallBack = (shadowRoot: ShadowRoot) => void

the third one has been removed but we still have three different concepts with similar names.

IMO:

- `interface ShadowDomCallBacks` could be `ShadowRootController` since it exposes controls for a single shadow root
- `shadowDomCallBacks = new Map<ShadowRoot, ShadowDomCallBacks>()` would then be `controllerByShadowRoot = new Map<ShadowRoot, ShadowRootController>()`

For `ShadowDomCallBack = (shadowRoot: ShadowRoot) => void`, would it not more precise to name it `ShadowRootCallBack`?

Reply:
thank you for the explainations, The type is gone from the `record.ts` (unless I am missing something when grepping the code base)

> interface ShadowDomCallBacks could be ShadowRootController since it exposes controls for a single shadow root

_loosely held opinion:_ We would have have `ShadowRootController` & `ShadowRootsController`. Having a single `s` of difference make it less clear to me But I can rename it `ShadowRootCallBacks`. What do you think?


> For ShadowDomCallBack = (shadowRoot: ShadowRoot) => void, would it not more precise to name it ShadowRootCallBack?

Indeed it is.

Reply:
> The type is gone from the record.ts (unless I am missing something when grepping the code base)

all good!

> We would have have ShadowRootController & ShadowRootsController. Having a single s of difference make it less clear to me But I can rename it ShadowRootCallBacks. What do you think?

IMO, the two components have similar responsibilities (flush, stop) so having a similar naming for it (Controller) could help the understanding.
For the single `s` of difference, I'd say it should not be too problematic as long are we are not mixing too much the two in the same context, cf https://github.com/DataDog/browser-sdk/pull/1787#discussion_r1055446995, inlining `shadowRootsController` could remove some occurrences.